crates/bitwarden-crypto/src/xchacha20.rs

```diff
@@ -0,0 +1,137 @@
+//! # XChaCha20Poly1305 operations
+//!
+//! Contains low level XChaCha20Poly1305 operations used by the rest of the crate.
+//!
+//! In most cases you should use the [EncString][crate::EncString] with
+//! [KeyEncryptable][crate::KeyEncryptable] & [KeyDecryptable][crate::KeyDecryptable] instead.
+
+use chacha20poly1305::{AeadCore, AeadInPlace, KeyInit, XChaCha20Poly1305};
+use generic_array::{typenum::U24, GenericArray};
+
+/**
+ * Note:
+ * XChaCha20Poly1305 encrypts data, and authenticates both the cipher text and associated
+ * data. This does not provide key-commitment, and assumes there can only be one key.
+ *
+ * If multiple keys are possible, a key-committing cipher such as
+ * XChaCha20Poly1305Blake3CTX should be used: `https://github.com/bitwarden/sdk-internal/pull/41` to prevent invisible-salamander style attacks.
+ * `https://eprint.iacr.org/2019/016.pdf`
+ * `https://soatok.blog/2024/09/10/invisible-salamanders-are-not-what-you-think/`
+ */
+use crate::CryptoError;
+
+#[allow(unused)]
+pub(crate) fn generate_nonce() -> GenericArray<u8, U24> {
+    XChaCha20Poly1305::generate_nonce(rand::thread_rng())
+}
+
+#[allow(unused)]
+pub(crate) fn encrypt_xchacha20_poly1305(
+    nonce: &[u8; 24],
+    key: &[u8; 32],
+    plaintext_secret_data: &[u8],
+    associated_data: &[u8],
+) -> Result<Vec<u8>, CryptoError> {
+    // This buffer contains the plaintext, that will be encrypted in-place
+    let mut buffer = Vec::from(plaintext_secret_data);
+    XChaCha20Poly1305::new(GenericArray::from_slice(key))
+        .encrypt_in_place(
+            GenericArray::from_slice(nonce),
+            associated_data,
+            &mut buffer,
+        )
+        .map_err(|_| CryptoError::InvalidKey)?;
```

Comment:
What's the error case for chacha20 encryption? Key and nonce sizes are checked at compile time, and as far as I know, chacha20 doesn't have any specific key format requirements.

If the only error case is the buffer not having enough space, I think we should be fine calling `.expect("")` and not returning a Result.